### PR TITLE
ENH: Add pre-commit check for setup.cfg options.extras_require

### LINKED ISSUE #48949: ENH: pre-commit check for setup.cfg options.extras_require
### Feature Type

- [ ] Extending the existing CI checks for package versioning to account for optional extras now defined in `setup.cfg`

### Problem Description

After #47336 added optional_extras in `setup.cfg`, we do not have a programmatic check that versions are aligned across the all places they are specified in pandas.

@JMBurley & @mroeschke  active in prior thread.

### Feature Description

A pre-commit check that ensure that the min version in setup.cfg are aligned with other areas we specify min version (would mean augmenting `scripts/validate_min_versions_in_sync.py`)

### Alternative Solutions

N/A.  Simplest to extend existing actions to cover new usage.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by kostyafarber**
Augmented `scripts/validate_min_versions_in_sync.py` to ensure that minimum versions are aligned in other areas where minimum versions are specified (`setup.cfg`). 

- [x] closes #48949(Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by mroeschke — CHANGES_REQUESTED** [review]
Is the check working correctly as https://github.com/pandas-dev/pandas/actions/runs/3307040746/jobs/5458375266 is failing?

Moreover, it would be great if the diff could be made clearer to state which file states which version e.g.

```
{package}
{file_1}: {version_1}
{file_2}: {version_2}
{file_3}: {version_3}
```

**Comment by JMBurley** [comment]
Bringing over my main issue comment:

> Based on experience my key goal here would be **clear error messages to help users find the problem**, so I'd then compare the diff list against each the pkg list from each source file to identify which file is missing what pkg and print that as part of the failure message.

I agree with mroescke's suggestion for how to format that information.

**Comment by kostyafarber** [comment]
Yes I agree we can make the diff more informative, I will work on making those changes. However, not sure why the check isn't working as intended. 

Following from the logic of the previous iteration of `scripts/validate_min_versions_in_sync.py` which was taking the symmetric difference of the dependency dicts, we decided to go with the approach of subtracting the intersection of all dicts (adding the dependencies dict from `setup.cfg`) from the union i.e `(a | b | c) - (a & b & c)`. to obtain the diff.

My hunch is maybe it's picking up versions that are defined in one place but not anywhere else?

Maybe I'm missing something but will have a crack and look into it.

**Comment by kostyafarber** [comment]
Question: are we also including 'test' libraries in 'setup.cfg' in our checks? I saw earlier in the script 'pytest' and didn't know if we are extending the same in this case.

**Comment by mroeschke** [comment]
> However there are more libraries than just pytest in tests are we keeping those or dropping them too?

Maybe can ignore the testing group for now. They can be addressed in a follow up PR

**Comment by kostyafarber** [comment]
Update: I nearly have it working but there are a couple of things I wanted to raise:

- When a library is specified in one file but none of the others OR all of the files but one:

```
qtpy
/Users/kostyafarber/repos/pandas/ci/deps/actions-38-minimum_versions.yaml: Not specified
/Users/kostyafarber/repos/pandas/pandas/compat/_optional.py: Not specified
/Users/kostyafarber/repos/pandas/setup.cfg: 2.2.0
```

```
hypothesis
/Users/kostyafarber/repos/pandas/ci/deps/actions-38-minimum_versions.yaml: 6.13.0
/Users/kostyafarber/repos/pandas/pandas/compat/_optional.py: 6.13.0
/Users/kostyafarber/repos/pandas/setup.cfg: Not specified
```
Is this something we should be picking up? Or are these false positives.

- There are a couple of instances when the versions are the "same" but have different formatting of versions (leading zero for example):

```
fsspec
/Users/kostyafarber/repos/pandas/ci/deps/actions-38-minimum_versions.yaml: 2021.07.0
/Users/kostyafarber/repos/pandas/pandas/compat/_optional.py: 2021.07.0
/Users/kostyafarber/repos/pandas/setup.cfg: 2021.7.0
```

Not sure how we want to deal with those. Would probably just be easier to make sure these follow the same conventions?

**Comment by mroeschke** [comment]
Ideally all files should have the same library with the same version (and format) specified.

**Comment by kostyafarber** [comment]
Okay I'll commit my changes shortly. Looks like there are a few libraries not in sync (if the check is working correctly). The pre-commit hook will expose them and you can let me know if we want to change/update/add them in this PR.

**Comment by kostyafarber** [comment]
Is there anything else I need to do here?

**Comment by MarcoGorelli** [comment]
The check is failing at the moment, so the versions should be aligned to get the check to pass before this can be merged

It's probably worth also checking that any minimum aligns with with what's been reported as the minimum version in the whatsnew notes. For example, `bottleneck` is listed as `>=1.3.1` in `setup.cfg`, but `>= 1.3.2` in the 1.5.0 whatsnew note:

https://github.com/pandas-dev/pandas/blob/bcb8346e8106be4267ec77dfc603d0d77a3fda81/setup.cfg#L124

https://github.com/pandas-dev/pandas/blob/145e527d717ada8d4648ac3d4b22a07d07143a31/doc/source/whatsnew/v1.5.0.rst#L563

So, it'd be good to align the `bottleneck` minimum to `1.3.2` across these files

**Comment by kostyafarber** [comment]
Okay cool. I'll check with what's in the what's new and change accordingly.

**Comment by kostyafarber** [comment]
I've aligned versions across the files (checking what's in `whatsnew`) and the check should pass now.

**Comment by kostyafarber** [comment]
Some of the packages I've aligned seem to be causing in issue in: https://github.com/pandas-dev/pandas/actions/runs/3362264680/jobs/5573763359

```
Encountered problems while solving:
    - nothing provides requested pyqt5 5.15.1**
    - package s3fs-2021.8.0-pyhd8ed1ab_0 requires aiobotocore 1.4.0, but none of the providers can be installed
```

Not sure any of the other fails are to do with this PR.

File: ci/deps/actions-38-minimum_versions.yaml
Comment (mroeschke) [inline ci/deps/actions-38-minimum_versions.yaml:16]: `boto3` is just a testing dependency so I don't think it should have a min version. I think you can just remove this from `setup.cfg` and leave this here

**Comment by mroeschke** [comment]
I think for `pyqt5` it will need to be installed with `pip` like in the conda env yaml like

```
- pip
   - pyqt5=version
```

**Comment by JMBurley** [comment]
@mroeschke @kostyafarber Regarding the build error
`s3fs-2021.8.0-pyhd8ed1ab_0 requires aiobotocore 1.4.0`

This is a common issue.  `aiobotocore` is a huge packaging problem, and the fact that `s3fs>0.4.0` relies on aiobotocore is a short-sighted choice that creates problems across the python ecosystem. (in summary: you can't use `aiobotocore` alongside anything else that interacts with `boto3`.  Which is a LOT of stuff.  Anyone wanting to unit test a repo will be using `moto`, for instance.)

I was mentioning in back in June for Pandas [here](https://github.com/pandas-dev/pandas/issues/39164#issuecomment-1154526531), and it has been discussed more extensively in other repos (eg. [here](https://github.com/fsspec/s3fs/issues/357#issuecomment-964627471))

I am not sure is there is a functional reason for pandas to advance the s3fs version pin beyond 0.4.0.  My 2 cents are that it is a mistake and the s3fs version needs to be held at 0.4.0 until the aiobotocore issue is resolved.  I'm happy to make a separate issue, adr, or PDEP to have that discussion if it needs to be had (@mroeschke ?)

File: ci/deps/actions-38-minimum_versions.yaml
Comment (kostyafarber) [reply ci/deps/actions-38-minimum_versions.yaml:16]: Sure thing will do.

**Comment by kostyafarber** [comment]
@JMBurley sounds like a mess. I've pinned `s3fs` at 0.4.0 for now.

**Comment by kostyafarber** [comment]
Not sure why these tests are failing.

File: ci/deps/actions-38-minimum_versions.yaml
Comment (mroeschke) [inline ci/deps/actions-38-minimum_versions.yaml:60]: Why is this `>=`? This build we want to test the minimum versions exactly

File: ci/deps/actions-38-minimum_versions.yaml
Comment (mroeschke) [inline ci/deps/actions-38-minimum_versions.yaml:49]: This dependency was bumped in 1.5 to a newer version: https://pandas.pydata.org/pandas-docs/stable/whatsnew/v1.5.0.html#backwards-incompatible-api-changes

So we shouldn't be testing an older version

Comment (kostyafarber) [reply ci/deps/actions-38-minimum_versions.yaml:49]: This was causing build issues that's why I pinned it at 0.4.0 https://github.com/pandas-dev/pandas/pull/49261#issuecomment-1297496447.

What is the verdict on what version we want to put in here for `s3fs`?

File: ci/deps/actions-38-minimum_versions.yaml
Comment (kostyafarber) [reply ci/deps/actions-38-minimum_versions.yaml:60]: Typo will change it.

File: ci/deps/actions-38-minimum_versions.yaml
Comment (kostyafarber) [reply ci/deps/actions-38-minimum_versions.yaml:49]: @JMBurley cc for visibility

Comment (JMBurley) [reply ci/deps/actions-38-minimum_versions.yaml:49]: Thanks, @kostyafarber .  I'm not sure if pandas **can** be built successfully with s3fs>=0.4.0 (and other packages active that require `boto3`), and **I know that pandas cannot be packaged into larger production environments when it enforces** `s3fs>=0.4.0` because professional codebases need other `boto3` imports.

However, I'm not a pandas maintainer so I don't get to make an assertion on what to pin s3fs to.  @mroeschke  controls the initial take on this.  I'd make the pitch for not advancing the s3fs version pin, because I do not know of any benefits we need (could be a lack of awareness on my part) and I do know of major python ecosystem headaches from advancing the pin.

My team pins `s3fs<=0.4.0` versions internally for this exact reason.  Ditto my colleagues controlling data science at other companies.

**Comment by kostyafarber** [comment]
`PyQt5` is causing some import errors for one of the tests:

https://github.com/pandas-dev/pandas/actions/runs/3377701924/jobs/5606956935 at line `214`.

```
module = <module 'PyQt5' from '/home/runner/micromamba/envs/test/lib/python3.8/site-packages/PyQt5/__init__.py'>

    def get_version(module: types.ModuleType) -> str:
        version = getattr(module, "__version__", None)
        if version is None:
            # xlrd uses a capitalized attribute name
            version = getattr(module, "__VERSION__", None)
    
        if version is None:
            if module.__name__ == "brotli":
                # brotli doesn't contain attributes to confirm it's version
                return ""
            if module.__name__ == "snappy":
                # snappy doesn't contain attributes to confirm it's version
                # See https://github.com/andrix/python-snappy/pull/119
                return ""
>           raise ImportError(f"Can't determine version for {module.__name__}")
E           ImportError: Can't determine version for PyQt5
```

We cannot access the version for this library using `getattr(module, "__version__", None)`. 

Looks like the version is in the `pyproject.toml`. Do you have any ideas for this issue?

**Comment by MarcoGorelli** [comment]
Looks like they have `from PyQt5.QtCore import QT_VERSION_STR, PYQT_VERSION_STR`

**Comment by kostyafarber** [comment]
Seems like changing from `PyQt5` to 'pyqt5' in `_optional.py` fixed the import issue. Not sure how though.

**Comment by kostyafarber** [comment]
@mroeschke looks to be okay. Just one fail due to something with `matplotlib` but don't think it's related here.

**Comment by kostyafarber** [comment]
Gonna leave it here. If there's anything else I need to do let me know.

**Comment by kostyafarber** [comment]
Any updates on this one?

File: setup.cfg
Comment (MarcoGorelli) [inline setup.cfg:107]: nice - I think older versions of matplotlib were failing the CI checks, so it's good you've caught this

File: ci/deps/actions-38-minimum_versions.yaml
Comment (MarcoGorelli) [reply ci/deps/actions-38-minimum_versions.yaml:49]: Hi @JMBurley - could you please clarify about

>  I'm not sure if pandas can be built successfully with s3fs>=0.4.0 

? The `environment.yml` file has a later version of `s3fs`, and building pandas with the `environment.yml` file works fine 

https://github.com/pandas-dev/pandas/blob/0b931173c842a3476646b627422cf943d15288f6/environment.yml#L48

File: setup.cfg
Comment (kostyafarber) [reply setup.cfg:107]: No worries. Only caught it because of the script 😄

**Comment by JMBurley** [comment]
@MarcoGorelli 

The issue is well covered in prior links ([here](https://github.com/fsspec/s3fs/issues/357)) but in summary when s3fs required aiobotocore (which in turn required boto3) it was incompatible with any other package that used boto3 because aiobotocore has extremely tight OLD pins to boto3 that other libraries generally don't.

However, it looks like this may have been quietly resolved in the past few months as my simple test `.cfg` file no longer triggers a pip-compile error

```
[metadata]
name = s3fs-test
version = 1.0.0
author = JMBurley
classifiers =
    Programming Language :: Python :: 3.9
    Operating System :: OS Independent

[options]
zip_safe = False
include_package_data = True
# min python 3.8 for Walrus operator; preclude auto-updates to future python 4.0
python_requires= >=3.9, <4.
# In principle, should make list below accept a broad range of packages, but not new major versions (risk of
# breaking changes). the `~=` and `>=2.1, <3.0` syntax is useful for this
install_requires =
    boto3>=1.20.0
    moto>=3.0, <4.0  # [s3] manages boto3, botocore. Thorny mutual dependency problem w/ aiobotocore via s3fs.
    s3fs>0.4.0  # s3fs<0.5 doesn't need aiobotocore (and its strict botocore pinning).  Removes incompatibility with moto's botocore needs.
```
run with  `pip-compile --output-file=requirements_example.txt --resolver=backtracking`

In which case I remove my objection if we can get tests working here with a newer s3fs version.  Although I am concerned that aiobotocore has historically caused packaging issues frequently so might not be "safe" long-term.

**Comment by kostyafarber** [comment]
@JMBurley bumped `s3fs` awaiting green tick.

**Comment by kostyafarber** [comment]
> @JMBurley bumped `s3fs` awaiting green tick.

Looks good 👍🏻

File: scripts/validate_min_versions_in_sync.py
Comment (MarcoGorelli) [inline scripts/validate_min_versions_in_sync.py:63]: just in case some package with `pip` as part of its name were added to the dependencies
```suggestion
        elif "- pip:" in line:
```

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me - got a really minor comment but not a blocker

File: scripts/validate_min_versions_in_sync.py
Comment (kostyafarber) [reply scripts/validate_min_versions_in_sync.py:63]: Yes that's a good point. Thanks for that.

**Review by JMBurley — APPROVED** [review]
LGTM thanks for the work here!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Awesome work @kostyafarber; glad this caught some issues already :) and thanks @JMBurley for your continued guidance here

**Comment by kostyafarber** [comment]
No worries @mroeschke glad I could help. Thanks for the support everybody! 🚀

**Pull request merged**
